omod/src/main/java/org/openmrs/web/controller/PortletController.java

```diff
@@ -195,8 +202,38 @@ public ModelAndView handleRequest(HttpServletRequest request, HttpServletRespons
 					}
 					
 					if (Context.hasPrivilege(PrivilegeConstants.GET_OBS)) {
-						List<Obs> patientObs = Context.getObsService().getObservationsByPerson(p);
-						model.put("patientObs", patientObs);
+						// Get pagination parameters
+						Integer page = getPageParameter(request);
+						Integer pageSize = getPageSizeParameter(request, as);
+						
+						Person person = (Person) p;
+						List<Person> persons = Collections.singletonList(person);
+						
+						// Setup parameters for database-level pagination
+						List<Encounter> encounters = null;
+						List<Concept> questions = null;
+						List<Concept> answers = null;
+						List<PERSON_TYPE> personTypes = null;
+						List<Location> locations = null;
+						List<String> sort = Collections.singletonList("obsDatetime desc");
+						Integer obsGroupId = null;
+						Date fromDate = null;
+						Date toDate = null;
+						boolean includeVoided = false;
+						
+						// Get observations for the current page
+						List<Obs> paginatedObs = Context.getObsService().getObservations(persons, encounters, questions,
```

Comment:
Though the comment says get observations for the current page, i do not seen anything in the getObservations method call that does so. Was it just a typo?

Reply:
Ooh yeah, my bad the getObservations() method doesn't have parameters for offset/limit pagination, let me remove that comment

Reply:
So then, shall we then have any dashboard loading optimisation if the method loads all of them in memory?

Reply:
Okay, I critically looked at the ObsService interface, didn't notice that it lacks proper pagination parameters (offset/limit). Maybe this should be done as a DAO level implementation so we can have a DB level pagination support, but I would think of doing it this way as well i.e. we can use an existing API features like mostRecentN to limit initial load like so

```suggestion
List<Obs> paginatedObs = Context.getObsService().getObservations(
    persons,
    null,
    null,
    null,
    null,
    null,
    Collections.singletonList("obsDatetime desc"),
    pageSize,  // mostRecentN - only fetch what we need
    null,
    null,
    null,
    false
);
```

or using date filtering

```suggestion
Calendar cal = Calendar.getInstance();
cal.add(Calendar.DAY_OF_MONTH, -30);
Date thirtyDaysAgo = cal.getTime();

List<Obs> recentObs = Context.getObsService().getObservations(
    persons,
    null,
    null,
    null,
    null,
    null,
    sort,
    null,
    null,
    thirtyDaysAgo,
    new Date(),
    false
);
```

Your thoughts @dkayiwa

Reply:
Okay!
@eudson Any thoughts on this, is it acceptable to only show the most recent observations?